# HG changeset patch
# User Torsten <mttl@mailbox.org>
# Date 1544388171 -3600
#      Sun Dec 09 21:42:51 2018 +0100
# Node ID b23954fd0ea68fe1071968a8f1be96a2f2a967ca
# Parent  ef0998af70f9436cdd9cc3a0271dc0cba66ff786
fix encoding combo box for not supported default encoding (bug #55139)

* resource-manager.cc (do_combo_encoding): put name of codec for locale
  on top of selection combo box in case the default from the preferences
  is not supported by the current system

diff --git a/libgui/src/resource-manager.cc b/libgui/src/resource-manager.cc
--- a/libgui/src/resource-manager.cc
+++ b/libgui/src/resource-manager.cc
@@ -393,38 +393,47 @@ namespace octave
   // initialize a given combo box with available text encodings
   void resource_manager::do_combo_encoding (QComboBox *combo, QString current)
   {
     QStringList all_codecs;
     do_get_codecs (&all_codecs);
 
     // get the value from the settings file if no current encoding is given
     QString enc = current;
+
+    bool default_exists = false;
+    if (QTextCodec::codecForName (ed_default_enc.def.toString ().toLatin1 ()))
+      default_exists = true;
+
     if (enc.isEmpty ())
       {
         enc = m_settings->value (ed_default_enc.key, ed_default_enc.def).toString ();
+
         if (enc.isEmpty ())  // still empty?
-          enc = ed_default_enc.def.toString ();     // take default
-
-        if (! QTextCodec::codecForName (enc.toLatin1 ())) // does it exist?
-          enc = "";   // no, so clear it
+          {
+            if (default_exists)
+              enc = ed_default_enc.def.toString ();
+            else
+              enc = QTextCodec::codecForLocale ()->name ();
+          }
       }
 
     // fill the combo box
     foreach (QString c, all_codecs)
       combo->addItem (c);
 
-    // prepend the current/default item and select it
-    if (! enc.isEmpty ())
-      {
-        combo->insertSeparator (0);
-        combo->insertItem (0, ed_default_enc.def.toString ());
+    // prepend the default item
+    combo->insertSeparator (0);
+    if (default_exists)
+      combo->insertItem (0, ed_default_enc.def.toString ());
+    else
+      combo->insertItem (0, QTextCodec::codecForLocale ()->name ());
 
-        int idx = combo->findText (enc);
-        if (idx >= 0)
-          combo->setCurrentIndex (idx);
-        else
-          combo->setCurrentIndex (0);
-      }
+    // select the default or the current one
+    int idx = combo->findText (enc);
+    if (idx >= 0)
+      combo->setCurrentIndex (idx);
+    else
+      combo->setCurrentIndex (0);
 
     combo->setMaxVisibleItems (12);
   }
 }
